experanto/interpolators.py

```diff
@@ -140,113 +149,154 @@ def normalize_data(self, data):
+            self.end_time
+            + (np.min(self._phase_shifts) if len(self._phase_shifts) > 0 else 0),
+        )
 
-                idx_upper = idx_upper[compute_mask]
-                idx_lower = idx_lower[compute_mask]
+    def interpolate(self, times: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        valid = self.valid_times(times)
+        valid_times = times[valid]
 
-                times_lower = idx_lower * self.time_delta
-                times_upper = idx_upper * self.time_delta
-                denom = times_upper - times_lower
+        if len(valid_times) == 0:
+            return np.array([]), valid
 
-                times_valid = valid_times[compute_mask]
+        idx_lower = np.floor(
+            (
+                valid_times[:, np.newaxis]
+                - self._phase_shifts[np.newaxis, :]
+                - self.start_time
+            )
+            / self.time_delta
+        ).astype(int)
 
-                lower_signal_ratio = ((times_upper - times_valid) / denom)[:, None]
-                upper_signal_ratio = ((times_valid - times_lower) / denom)[:, None]
+        if self.interpolation_mode == "nearest_neighbor":
+            data = np.take_along_axis(self._data, idx_lower, axis=0)
+            return data, valid
 
-                data_lower = self._data[idx_lower]
-                data_upper = self._data[idx_upper]
+        elif self.interpolation_mode == "linear":
+            idx_upper = idx_lower + 1
+            overflow_mask = idx_upper >= self._data.shape[0]
 
-                interpolated = np.full(
-                    (valid_times.shape[0], data_lower.shape[1]), np.nan
-                )
-                interpolated[compute_mask] = (
-                    lower_signal_ratio * data_lower + upper_signal_ratio * data_upper
+            if np.any(idx_lower < 0):  # should not be possible
+                warnings.warn(
+                    f"Interpolation index {idx_lower} is negative. This should not happen."
                 )
+                overflow_mask = overflow_mask | idx_lower < 0
+
+            times_lower = idx_lower * self.time_delta
+            times_upper = idx_upper * self.time_delta
+            denom = times_upper - times_lower
+
+            time_dim = valid_times[:, np.newaxis] - self._phase_shifts[np.newaxis, :]
+
+            lower_numerator = times_upper - time_dim
+            upper_numerator = time_dim - times_lower
+
+            lower_signal_ratio = lower_numerator / denom
+            upper_signal_ratio = upper_numerator / denom
+
+            _, cols = np.indices(idx_lower.shape)
+            data_lower = self._data[idx_lower, cols]
+            data_upper = self._data[idx_upper, cols]
+
+            interpolated = (
+                lower_signal_ratio * data_lower + upper_signal_ratio * data_upper
+            )
 
-                valid = valid[~overflow_mask]
+            valid = valid[~overflow_mask.any(axis=1)]
```

Comment:
why is it `.any(axis=1)` here?

Reply:
When we have phase shifts, overflow can happen for some neurons (with a high phase shift) and not for others (with a low phase shift). So we check if an overflow is happening for any neuron at a certain time and disregard the time then completely.

When we don't have phase shifts, the times are the same for every neuron, leading to one dimension less and therefore `.any(axis=1)` is not necessary.

Reply:
could we please keep this conversation open for now, I need to think a bit about it